File: extensions/coding-workflow/skills/design-nfr/scripts/check_nfr.py
#!/usr/bin/env python3
"""
design-nfr（④非功能性设计）硬规则机器验证

Usage:
    python3 check_nfr.py <topic_dir>

检查项：
  ①结构性：non-functional-design.md 存在 / verdict:pass / 关键章节 / 无占位符 / review-nfr APPROVED
  ②引用：
    - 缓解项回灌登记表每行有「验收方式」列且值 ∈ {代码测试, 骨架约束, 运维项}
    - 代码测试类的缓解项有对应 NFR-AC（归属 UC + 断言）
    - 无 ❌（不可接受）项残留——如有说明未回 Step 3 重选方案

Exit code: 0 = 全过，1 = 有硬伤
"""

import os
import re
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
_LIB_DIR = os.path.join(
    os.path.dirname(os.path.abspath(__file__)),
    os.pardir, os.pardir, "design-clarity", "scripts",
)
sys.path.insert(0, _LIB_DIR)
from _shared_check_lib import (
    CheckReport, resolve_topic_dir, read_text,
    check_file_exists, check_frontmatter_verdict, check_required_sections,
    check_no_placeholders, check_review_verdict, has_heading, extract_section,
)

DELIVERABLE = "non-functional-design.md"
VALID_ACCEPTANCE = {"代码测试", "骨架约束", "运维项"}


def main():
    topic_dir = resolve_topic_dir()
    report = CheckReport("nfr")
    md_path = os.path.join(topic_dir, DELIVERABLE)

    # ① 结构性
    if not check_file_exists(report, f"{DELIVERABLE} 存在", md_path):
        report.finalize_and_exit(topic_dir)
        return
    check_frontmatter_verdict(report, md_path, "pass")
    check_required_sections(
        report, md_path, "关键章节",
        [r"分析矩阵|风险矩阵", r"缓解项回灌|Mitigation"],
    )
    check_no_placeholders(report, "无占位符", md_path)
    check_review_verdict(report, topic_dir, "nfr", "APPROVED")

    # ② 缓解项回灌表每行有验收方式列且值合法
    mitigation_section = extract_section(md_path, r"缓解项回灌|Mitigation")
    if mitigation_section:
        # 提取表格行（含「验收方式」列）
        table_rows = [l for l in mitigation_section.splitlines()
                      if l.strip().startswith("|") and "----" not in l and "验收方式" not in l]
        if table_rows:
            invalid_acceptance = []
            for row in table_rows:
                cells = [c.strip() for c in row.split("|")[1:-1]]
                # 找验收方式列（倒数第2列，状态是最后1列）
                if len(cells) >= 2:
                    acceptance = cells[-2] if cells[-1] in ("待落", "已落", "PASS") else cells[-2]
                    # 宽松匹配：单元格含任一合法值
                    matched = [v for v in VALID_ACCEPTANCE if v in acceptance]
                    if not matched:
                        invalid_acceptance.append(f"'{acceptance}'")
            if invalid_acceptance:
                report.add_fail(
                    "验收方式列合法",
                    f"{len(invalid_acceptance)} 行验收方式不合法: {invalid_acceptance[:3]}（应 ∈ {VALID_ACCEPTANCE}）",
                )
            else:
                report.add_pass("验收方式列合法", f"{len(table_rows)} 行缓解项均标了合法验收方式")
        else:
            report.add_fail("缓解项回灌表", "缓解项回灌章节无表格行")
    else:
        report.add_fail("缓解项回灌表", "无「缓解项回灌登记」章节（MANDATORY）")

    # ② 无 ❌（不可接受）项残留
    content = read_text(md_path)
    # ❌ 在分析矩阵/详细分析中代表"不可接受需回退"，定稿不应残留
    unacceptable = re.findall(r"❌[^\n]*", content)
    # 过滤掉说明性文字里的 ❌（如"无 ❌ 项"）
    real_unacceptable = [u for u in unacceptable if "无" not in u[:3] and "不残留" not in u]
    if real_unacceptable:
        report.add_fail(
            "无 ❌ 不可接受项",
            f"残留 {len(real_unacceptable)} 处 ❌（不可接受项应已回 Step 3 重选方案）",
        )
    else:
        report.add_pass("无 ❌ 不可接受项", "无不可接受项残留")

    report.finalize_and_exit(topic_dir)


if __name__ == "__main__":
    main()

File: extensions/coding-workflow/src/cw/checks/check-nfr.ts
/**
 * full-nfr（④非功能性设计）硬规则机器验证。
 *
 * 移植自 skills/full-nfr/scripts/check_nfr.py。
 * 入参：topicDir（含 non-functional-design.md + changes/review-nfr.md + issues.md）。
 *
 * 检查项：
 *   ①结构性：交付物存在 / verdict:pass / 关键章节 / 无占位符 / review-nfr APPROVED
 *   ②引用：
 *     - 缓解项回灌表每行有「验收方式」列且值 ∈ {代码测试, 骨架约束, 性能混沌, 运维项}
 *     - 无 ❌（不可接受）项残留
 *     - 回灌表「回灌去向=③issue」的行，#N 真实存在于 issues.md（PHANTOM 形式检查）
 */

import { existsSync } from "node:fs";
import { join } from "node:path";

import { extractRealIssueIds } from "./check-issues.js";
import {
  checkFileExists,
  checkFrontmatterVerdict,
  checkNoPlaceholders,
  type CheckOutput,
  CheckReport,
  checkRequiredSections,
  checkReviewVerdict,
  extractSection,
  readText,
} from "./shared.js";

const DELIVERABLE = "non-functional-design.md";
const VALID_ACCEPTANCE = new Set(["代码测试", "骨架约束", "性能混沌", "运维项"]);
/** 报错清单截断（避免单条报错过长）。 */
const ERR_LIST_MAX = 5;

export function runCheckNfr(topicDir: string): CheckOutput {
  const report = new CheckReport("nfr");
  const mdPath = join(topicDir, DELIVERABLE);

  // ① 结构性
  if (!checkFileExists(report, `${DELIVERABLE} 存在`, mdPath)) {
    return report.toOutput({ writeReport: true, topicDir });
  }
  checkFrontmatterVerdict(report, mdPath, "pass");
  checkRequiredSections(report, mdPath, "关键章节", [
    "分析矩阵|风险矩阵",
    "缓解项回灌|Mitigation",
  ]);
  checkNoPlaceholders(report, "无占位符", mdPath);
  checkReviewVerdict(report, topicDir, "nfr", "APPROVED");

  // ② 缓解项回灌表每行有验收方式列且值合法
  const mitigationSection = extractSection(mdPath, "缓解项回灌|Mitigation");
  if (mitigationSection) {
    // 提取表格行（含「验收方式」列，跳过表头和分隔行）
    const tableRows = mitigationSection
      .split(/\r?\n/)
      .filter((l) => l.trim().startsWith("|") && !l.includes("----") && !l.includes("验收方式"));
    if (tableRows.length > 0) {
      const invalidAcceptance: string[] = [];
      for (const row of tableRows) {
        const cells = row.split("|").slice(1, -1).map((c) => c.trim());
        // 找验收方式列（倒数第2列）
        const MIN_NFR_TABLE_CELLS = 2;
        if (cells.length >= MIN_NFR_TABLE_CELLS) {
          const acceptance = cells[cells.length - MIN_NFR_TABLE_CELLS] ?? "";
          const matched = [...VALID_ACCEPTANCE].filter((v) => acceptance.includes(v));
          if (matched.length === 0) {
            invalidAcceptance.push(`'${acceptance}'`);
          }
        }
      }
      if (invalidAcceptance.length > 0) {
        report.addFail(
          "验收方式列合法",
          `${invalidAcceptance.length} 行验收方式不合法: ${JSON.stringify(invalidAcceptance.slice(0, ERR_LIST_MAX))}` +
            `（应 ∈ ${JSON.stringify([...VALID_ACCEPTANCE])}）`,
        );
      } else {
        report.addPass("验收方式列合法", `${tableRows.length} 行缓解项均标了合法验收方式`);
      }
    } else {
      report.addFail("缓解项回灌表", "缓解项回灌章节无表格行");
    }
  } else {
    report.addFail("缓解项回灌表", "无「缓解项回灌登记」章节（MANDATORY）");
  }

  // ② 无 ❌（不可接受）项残留
  //
  // nfr 模板自身含合法 ❌ 字符（B1 修复）：
  //   - 图例行：`（✅ 无风险 / ⚠️ 有风险已缓解 / ❌ 不可接受需回退 / — 不适用+理由）`
  //   - 回灌指针表说明行：`❌ ⑤还没写，查不了` / `❌ ⑥还没编排，查不了`（延期承诺标记）
  // 全文 grep ❌ 会误伤这些合法说明行，导致 agent 保留模板图例就 FAIL。
  //
  // 过滤策略（按行扫描，跳过说明性 ❌）：
  //   1. 图例行：同一行同时含 ✅ 和 ⚠️（4 个状态符号并列）→ 跳过
  //   2. 回灌指针表延期承诺说明：含「还没」/「查不了」→ 跳过
  //   3. 元描述（描述 ❌ 自身含义）：含「无风险」/「有风险已缓解」/「不适用」/
  //      「不残留」/「不可接受需回退」→ 跳过
  //   4. 兜底原过滤：「无 ❌」前缀（如"无 ❌ 项"）→ 跳过
  const content = readText(mdPath);
  const realUnacceptable: string[] = [];
  for (const rawLine of content.split(/\r?\n/)) {
    if (!rawLine.includes("❌")) continue;
    // 图例行（4 状态符号并列）
    if (rawLine.includes("✅") && rawLine.includes("⚠️")) continue;
    // 回灌指针表延期承诺说明
    if (rawLine.includes("还没") || rawLine.includes("查不了")) continue;
    // 元描述：解释 ❌ 含义的说明文字
    if (
      rawLine.includes("无风险") ||
      rawLine.includes("有风险已缓解") ||
      rawLine.includes("不适用") ||
      rawLine.includes("不残留") ||
      rawLine.includes("不可接受需回退")
    ) {
      continue;
    }
    // 兜底原过滤：「无 ❌」前缀（如"无 ❌ 项"）
    // 「无 ❌」前缀检测窗口（向前看 3 字符容纳「无 ❌」及混合空白）
    const PREFIX_WINDOW = 3;
    const idx = rawLine.indexOf("❌");
    const before = rawLine.slice(Math.max(0, idx - PREFIX_WINDOW), idx);
    if (before.includes("无")) continue;
    realUnacceptable.push(rawLine.trim());
  }
  if (realUnacceptable.length > 0) {
    report.addFail(
      "无 ❌ 不可接受项",
      `残留 ${realUnacceptable.length} 处 ❌（不可接受项应已回 Step 3 重选方案）: ${JSON.stringify(realUnacceptable.slice(0, ERR_LIST_MAX))}`,
    );
  } else {
    report.addPass("无 ❌ 不可接受项", "无不可接受项残留");
  }

  // ③ 回灌指针 PHANTOM 形式检查
  if (mitigationSection) {
    checkBackfeedPhantom(report, topicDir, mitigationSection);
  }

  return report.toOutput({ writeReport: true, topicDir });
}

/**
 * 回灌指针 PHANTOM 形式检查。
 *
 * 回灌表「回灌去向=③issue」的行，提取 #N，核对是否在 issues.md 的 issue 定义里真实存在。
 * 机器只查 PHANTOM（指针断裂）；MISMATCH（属性不符）+ ORPHAN 靠 Step2 重建器。
 * issues.md 不存在时降级为 SKIP。
 */
function checkBackfeedPhantom(
  report: CheckReport,
  topicDir: string,
  mitigationSection: string,
): void {
  const issuesPath = join(topicDir, "issues.md");
  if (!existsSync(issuesPath)) {
    report.addSkip(
      "回灌③指针 PHANTOM",
      "issues.md 不存在，跳过（MISMATCH/ORPHAN 仍由 Step2 重建器查）",
    );
    return;
  }

  const issuesContent = readText(issuesPath);
  const realIssueIds = new Set(extractRealIssueIds(issuesContent));

  // 从回灌表提取「回灌去向」含 ③/issue 的行里的 #N 引用
  const tableRows = mitigationSection
    .split(/\r?\n/)
    .filter((l) => l.trim().startsWith("|") && !l.includes("----"));
  const phantomRefs: string[] = [];
  let checked = 0;
  for (const row of tableRows) {
    const cells = row.split("|").slice(1, -1).map((c) => c.trim());
    // 跳过「回灌去向」表头行
    if (cells.some((c) => c.includes("回灌去向") || c.includes("去向"))) continue;
    // 只看声明指向 ③ 的行（回灌去向列含「③」或「issue」且含 #N）
    const targets3 = cells.filter(
      (c) => (c.includes("③") || c.toLowerCase().includes("issue")) && /#\d+/.test(c),
    );
    for (const cell of targets3) {
      for (const m of cell.matchAll(/#(\d+)/g)) {
        checked++;
        const ref = m[1]!;
        if (!realIssueIds.has(ref)) {
          phantomRefs.push(`#${ref}`);
        }
      }
    }
  }

  if (phantomRefs.length > 0) {
    report.addFail(
      "回灌③指针 PHANTOM",
      `${phantomRefs.length} 处回灌指针指向不存在的 issue: ${JSON.stringify(phantomRefs.slice(0, ERR_LIST_MAX))}（issues.md 无此编号）`,
    );
  } else if (checked > 0) {
    report.addPass("回灌③指针 PHANTOM", `${checked} 处回灌③指针均指向真实存在的 issue`);
  } else {
    report.addSkip("回灌③指针 PHANTOM", "回灌表无指向 ③issue 的行（可能全去⑤/运维项）");
  }
}
